File: src/main/java/lechuck/intellij/vars/VariablesDialog.java
package lechuck.intellij.vars;

import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.table.JBTable;
import com.intellij.ui.table.TableView;
import com.intellij.util.EnvironmentUtil;
import com.intellij.util.ui.ListTableModel;
import net.miginfocom.swing.MigLayout;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.table.TableCellRenderer;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class VariablesDialog extends DialogWrapper {
    private final VariablesTextFieldWithBrowseButton myParent;
    private final VariablesTable myVarTable;
    private final JPanel myPanel;

    protected VariablesDialog(VariablesTextFieldWithBrowseButton parent) {
        super(parent, true);
        myParent = parent;
        Map<String, String> varMap = new LinkedHashMap<>(myParent.getVars());

        List<Variable> varList = VariablesTextFieldWithBrowseButton.convertToVariables(varMap);
        myVarTable = new MyVariablesTable(varList);

        JLabel label = new JLabel("Variables:");
        label.setLabelFor(myVarTable.getTableView().getComponent());

        myPanel = new JPanel(new MigLayout("fill, ins 0, gap 0, hidemode 3"));
        myPanel.add(label, "hmax pref, wrap");
        myPanel.add(myVarTable.getComponent(), "push, grow, wrap, gaptop 5");

        setTitle("Variables");
        init();
    }

    @Nullable
    @Override
    protected String getDimensionServiceKey() {
        return "VariablesDialog";
    }

    @NotNull
    @Override
    protected JComponent createCenterPanel() {
        return myPanel;
    }

    @Nullable
    @Override
    protected ValidationInfo doValidate() {
        for (Variable variable : myVarTable.getVariables()) {
            String name = variable.getName();
            String value = variable.getValue();
            if (StringUtil.isEmpty(name) && StringUtil.isEmpty(value)) {
                continue;
            }

            if (!EnvironmentUtil.isValidName(name)) {
                return new ValidationInfo("Invalid variable name: " + name);
            }
            if (!EnvironmentUtil.isValidValue(value)) {
                return new ValidationInfo("Invalid variable value: " + name + " = " + value);
            }
        }
        return super.doValidate();
    }

    @Override
    protected void doOKAction() {
        myVarTable.stopEditing();
        final Map<String, String> vars = new LinkedHashMap<>();
        for (Variable variable : myVarTable.getVariables()) {
            if (StringUtil.isEmpty(variable.getName()) && StringUtil.isEmpty(variable.getValue())) {
                continue;
            }
            vars.put(variable.getName(), variable.getValue());
        }
        myParent.setVars(vars);
        super.doOKAction();
    }

    private static class MyVariablesTable extends VariablesTable {
        MyVariablesTable(List<Variable> list) {
            TableView<Variable> tableView = getTableView();
            tableView.setVisibleRowCount(JBTable.PREFERRED_SCROLLABLE_VIEWPORT_HEIGHT_IN_ROWS);
            setValues(list);
            setPasteActionEnabled(true);
        }

        @Override
        protected ListTableModel createListModel() {
            return new ListTableModel(new MyNameColumnInfo(), new MyValueColumnInfo());
        }

        protected class MyNameColumnInfo extends NameColumnInfo {
            @Override
            public TableCellRenderer getCustomizedRenderer(Variable o, TableCellRenderer renderer) {
                return renderer;
            }
        }

        protected class MyValueColumnInfo extends ValueColumnInfo {
            @Override
            public boolean isCellEditable(Variable variable) {
                return true;
            }

            @Override
            public TableCellRenderer getCustomizedRenderer(Variable o, TableCellRenderer renderer) {
                return renderer;
            }
        }
    }
}

File: src/main/kotlin/lechuck/intellij/vars/VariablesDialog.kt
package lechuck.intellij.vars

import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.ui.table.JBTable
import com.intellij.util.ui.ListTableModel
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.table.TableCellRenderer
import net.miginfocom.swing.MigLayout

class VariablesDialog(private val parent: VariablesTextFieldWithBrowseButton) :
    DialogWrapper(parent, true) {

    private val varTable: VariablesTable
    private val panel: JPanel

    init {
        val varMap = LinkedHashMap(parent.getVars())

        val varList = VariablesTextFieldWithBrowseButton.convertToVariables(varMap)
        varTable = MyVariablesTable(varList)

        val label = JLabel("Variables:")
        label.labelFor = varTable.tableView.component

        panel =
            JPanel(MigLayout("fill, ins 0, gap 0, hidemode 3")).apply {
                add(label, "hmax pref, wrap")
                add(varTable.component, "push, grow, wrap, gaptop 5")
            }

        title = "Variables"
        init()
    }

    override fun getDimensionServiceKey(): String {
        return "VariablesDialog"
    }

    override fun createCenterPanel(): JComponent {
        return panel
    }

    override fun doValidate(): ValidationInfo? {
        val variables = varTable.getVariables()
        for (variable in variables) {
            val error = validationError(variable.name, variable.value)
            if (error != null) {
                return ValidationInfo(error)
            }
        }
        for (variable in variables) {
            val warning = validationWarning(variable.name)
            if (warning != null) {
                // non-blocking: names like these were accepted before this check existed, so
                // they must not stop an existing configuration from being saved
                return ValidationInfo(warning).asWarning().withOKEnabled()
            }
        }
        return super.doValidate()
    }

    override fun doOKAction() {
        varTable.stopEditing()
        val vars = LinkedHashMap<String, String>()
        for (variable in varTable.getVariables()) {
            if (variable.name.isEmpty() && variable.value.isEmpty()) {
                continue
            }
            vars[variable.name] = variable.value
        }
        parent.setVars(vars)
        super.doOKAction()
    }

    companion object {
        // what a Taskfile can reference as {{.NAME}}: Go template field syntax, which permits
        // Unicode letters and digits plus '_' (text/template's isAlphaNumeric), not just ASCII
        private val TEMPLATE_REFERENCABLE = Regex("[\\p{L}_][\\p{L}\\p{Nd}_]*")

        /** @return the message to show, or null if the variable is acceptable */
        internal fun validationError(name: String, value: String): String? {
            if (name.isEmpty() && value.isEmpty()) {
                return null
            }
            // These are task variables, not environment variables: they travel as one
            // 'NAME=VALUE' command-line argument that task splits on its first '='. An empty
            // name or a '=' anywhere in one silently yields a differently-named variable --
            // the environment-variable rules used before even allowed a leading '=' on Windows.
            if (name.isEmpty()) {
                return "Variable name cannot be empty"
            }
            if (name.contains('=')) {
                return "Variable name cannot contain '='"
            }
            // exec silently truncates an argument at a NUL byte
            if (name.contains('\u0000') || value.contains('\u0000')) {
                return "Variable cannot contain a NUL character"
            }
            // task interpolates variables into the shell command, so a line break in one ends the
            // command and the rest is run as another: 'task NAME=a<newline>b' fails with exit 127
            if (containsLineBreak(name)) {
                return "Variable name cannot contain a line break"
            }
            if (containsLineBreak(value)) {
                return "Variable value cannot contain a line break: $name"
            }
            return null
        }

        /**
         * @return a non-blocking message for a name that task accepts on the command line but a
         *   Taskfile cannot reference as a template field, or null if the name is fine
         */
        internal fun validationWarning(name: String): String? {
            if (name.isEmpty() || name.matches(TEMPLATE_REFERENCABLE)) {
                return null
            }
            return "'$name' cannot be referenced as {{.$name}} in a Taskfile"
        }

        private fun containsLineBreak(s: String) = s.contains('\n') || s.contains('\r')
    }

    private class MyVariablesTable(list: List<Variable>) : VariablesTable() {
        init {
            val tableView = tableView
            tableView.visibleRowCount = JBTable.PREFERRED_SCROLLABLE_VIEWPORT_HEIGHT_IN_ROWS
            setValues(list)
            setPasteActionEnabled(true)
        }

        override fun createListModel(): ListTableModel<Variable> {
            return ListTableModel(MyNameColumnInfo(), MyValueColumnInfo())
        }

        inner class MyNameColumnInfo : NameColumnInfo() {
            override fun getCustomizedRenderer(
                o: Variable?,
                renderer: TableCellRenderer?,
            ): TableCellRenderer? {
                return renderer
            }
        }

        inner class MyValueColumnInfo : ValueColumnInfo() {
            override fun isCellEditable(variable: Variable): Boolean {
                return true
            }

            override fun getCustomizedRenderer(
                o: Variable?,
                renderer: TableCellRenderer?,
            ): TableCellRenderer? {
                return renderer
            }
        }
    }
}
